constraint-solver/src/constraint_system.rs

```diff
@@ -150,6 +143,20 @@ impl<T, V> AlgebraicConstraint<GroupedExpression<T, V>> {
     }
 }
 
+impl<T: RuntimeConstant + Substitutable<V>, V: Clone + Eq + Ord>
+    AlgebraicConstraint<GroupedExpression<T, V>>
+{
+    /// Substitute a variable by a symbolically known expression. The variable can be known or unknown.
+    /// If it was already known, it will be substituted in the known expressions.
+    pub fn substitute_by_known(&mut self, variable: &V, substitution: &T) {
+        self.expression.substitute_by_known(variable, substitution);
+    }
+
+    pub fn degree(&self) -> usize {
+        self.expression.degree()
+    }
+}
+
 impl<V: Display> Display for AlgebraicConstraint<V> {
     fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
         write!(f, "{} = 0", self.expression)
```

Comment:
would be nice if we could use `GroupedExpression::to_signed_string` to extract the outermost negation. We could also convert to AlgebraicExpression and check if the outermost layer is a binary negation and split between lhs and rhs.